src/utils/__tests__/rate-limiter.test.ts

```diff
@@ -0,0 +1,58 @@
+import { RateLimiter } from '../rate-limiter';
+
+describe('RateLimiter', () => {
+  it('should not allow more than the rate limit to run concurrently', async () => {
+    const limit = new RateLimiter(3);
+    let activeTasks = 0;
+    const increaseActive = () => {
+      activeTasks++;
+    };
+    const decreaseActive = () => {
+      activeTasks--; limit.free();
+    };
+    const tasks = Array(5)
+      .fill(null).map(async () => {
+        await limit.allocate();
+
+        increaseActive();
+        expect(activeTasks).toBeLessThanOrEqual(3);
+
+        // Mock task duration (randomness to affect completion order)
+        await new Promise(resolve =>
+          setTimeout(resolve, 500 + 500 * Math.random())
+        );
+        decreaseActive();
+      });
+      await Promise.all(tasks);
+  });
+
+  it('should queue tasks and execute them in order', async () => {
+    const limit = new RateLimiter(2);
```

Comment:
i guess the issue in the test is here, resolving 2 promises at the same time never makes it execute in order, this should be 1